docs/running_lbann.rst

```diff
@@ -1,96 +1,434 @@
+          future.
+
+Data readers are responsible for managing a data set and providing
+data samples to models. A data set is comprised of independent data
+samples, each of which is made up of multiple tensors. For example, a
+data sample for a labeled image classification problem consists of an
+image tensor and a one-hot label vector.
+
+.. note:: The data readers are currently hard-coded to assume this
+          simple classification paradigm. Hacks are needed if your
+          data does not match it exactly, e.g. if a data sample is
+          comprised of more than two tensors. The most basic approach
+          is to flatten all tensors and concatenate them into one
+          large vector. The model is then responsible for slicing this
+          vector into the appropriate chunks and resizing the chunks
+          into the appropriate dimensions. Done correctly, this should
+          not impose any additional overhead.
+
+Specifically, data readers and models interact via input layers. Each
+model must have exactly one input layer and its output tensors are
+populated by a data reader every mini-batch step. This is typically
+performed by a background thread pool, so data ingestion will
+efficiently overlap with other computation, especially if the data
+reader's work is IO-bound or if the computation is largely on GPUs.
+
+.. note:: An input layer has an output tensor for each data sample
+          tensor. Since each data sample has two tensors (one for the
+          data and one for the label), it follows that every input
+          layer should have two child layers. To make parent/child
+          relationships unambiguous, we recommend manually creating
+          identity layers as children of the input layer.
+
+Note that layers within a model treat the data for a mini-batch as a
+single tensor where the leading dimension is the mini-batch
+size. Thus, corresponding tensors in all data samples must have the
+same dimensions. The data dimensions must be known from the beginning
+of the experiment and can not change. However, real data is rarely so
+consistent and some preprocessing is typically required.
+
+.. note:: `A major refactor of the preprocessing pipeline
+          <https://github.com/LLNL/lbann/pull/1014>`_ is pending. This
+          documentation will be updated once it is merged and the
+          interface stabilized.
+
+------------------------------------------------
+Python frontend
+------------------------------------------------
+
+LBANN provides a Python frontend with syntax reminiscent of `PyTorch
+<https://pytorch.org/>`_. See the `model zoo implementation of LeNet
+<https://github.com/LLNL/lbann/blob/develop/model_zoo/vision/lenet.py>`_
+for a simple example.
+
+Comments:
+
++ Under-the-hood, the Python frontend is actually a convenience
+  wrapper around the Protobuf frontend. The core infrastructure allows
+  users to configure an experiment, "compiles" it to a Prototext text
```

Comment:
Tense is a bit strange here. The verb tenses of "allows", "compiles", and "feeds" make sense but the presence of "configure" kind of throws that pattern off. 

Maybe 'and then "compiles" it to a Prototext file and feeds that into the Protobuf frontend.'